---
title: 'Figure 1 Theory Meets Figure 2 Experiments in the Study of Gene Expression'
date: '2019-05-10'
authors:
  - name: Rob Phillips
    corr: true
  - name: Nathan M. Belliveau
  - name: Griffin Chure
  - name: Hernan G. Garcia
  - name: Manuel Razo-Majia
  - name: Clarissa Scholes 
journal: 'Annual Reviews of Biophysics'
tags: ['gene regulation', 'statistical mechanics']
DOI: '10.1146/annurev-biophys-052118-115525'
links:
  - name: PDF
    url: 'Phillips2019.pdf'
  - name: Publisher
    url: 'https://www.annualreviews.org/content/journals/10.1146/annurev-biophys-052118-115525'
abstract: >
    It is tempting to believe that we now own the genome. The ability to read and rewrite it at will has ushered in a stunning period in the history of science. Nonetheless, there is an Achilles’ heel exposed by all of the genomic data that has accrued: We still do not know how to interpret them. Many genes are subject to sophisticated programs of transcriptional regulation, mediated by DNA sequences that harbor binding sites for transcription factors, which can up- or down-regulate gene expression depending upon environmental conditions. This gives rise to an input–output function describing how the level of expression depends upon the parameters of the regulated gene—for instance, on the number and type of binding sites in its regulatory sequence. In recent years, the ability to make precision measurements of expression, coupled with the ability to make increasingly sophisticated theoretical predictions, has enabled an explicit dialogue between theory and experiment that holds the promise of covering this genomic Achilles’ heel. The goal is to reach a predictive understanding of transcriptional regulation that makes it possible to calculate gene expression levels from DNA regulatory sequence. This review focuses on the canonical simple repression motif to ask how well the models that have been used to characterize it actually work. We consider a hierarchy of increasingly sophisticated experiments in which the minimal parameter set learned at one level is applied to make quantitative predictions at the next. We show that these careful quantitative dissections provide a template for a predictive understanding of the many more complex regulatory arrangements found across all domains of life.
---
